src/inmanta/server/agentmanager.py

```diff
@@ -264,6 +276,32 @@ async def _unpause_agent(
                 env.id, endpoints_with_new_primary, now=datetime.now().astimezone(), connection=connection
             )
 
+    async def _keep_agent_paused_on_resume(
+        self, env: data.Environment, endpoint: Optional[str] = None, connection: Optional[asyncpg.connection.Connection] = None
+    ) -> None:
+        """
+        Set the unpause_on_resume field of an agent (or all agents in an environment when the endpoint is set to None)
+        so that the agent is still paused after the environment is resumed
+        """
+
+        async with self.session_lock:
```

Comment:
`self.session_lock` exists to prevent concurrent access on data structures that hold sessions, like `self.sessions`. This code doesn't modify that data structure, so there is no need to acquire the lock.